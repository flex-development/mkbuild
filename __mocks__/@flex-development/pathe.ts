/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const format = vi.fn(actual.format)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)

export default {
  basename,
  dirname,
  extname,
  format,
  join,
  parse,
  relative,
  resolve
}
