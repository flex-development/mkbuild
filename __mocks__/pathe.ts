/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

import path, { type FormatInputPathObject } from 'node:path'

/**
 * Return the last portion of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} File name
 */
export const basename = vi.fn((p: string): string => path.basename(p))

/**
 * Returns the directory name of `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Directory containing `p`
 */
export const dirname = vi.fn((p: string): string => path.dirname(p))

/**
 * Returns the extension of `p`, from the last `'.'` to end of string in the
 * last portion of the path.
 *
 * If there is no `'.'` in the last portion of the path or the first character
 * of it is `'.'`, an empty string will be returned.
 *
 * @param {string} p - Path to evaluate
 * @return {string} Extension of `p`
 */
export const extname = vi.fn((p: string) => path.extname(p))

/**
 * Returns a path string from an object - the opposite of {@link parse}.
 *
 * @param {FormatInputPathObject} p - Path object to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const format = vi.fn((p: FormatInputPathObject) => path.format(p))

/**
 * Parses `p`.
 *
 * @param {string} p - Path to evaluate
 * @return {path.ParsedPath} Object representing `p` parsed
 */
export const parse = vi.fn((p: string) => path.parse(p))

/**
 * Returns the relative path from `from` to `to` based on the current working
 * directory.
 *
 * @param {string} from - Start path
 * @param {string} to - Destination path
 * @return {string} Relative path from `from` to `to`
 */
export const relative = vi.fn((from: string, to: string): string => {
  return path.relative(from, to)
})

/**
 * Starting from leftmost {from} parameter, resolves {to} to an absolute path.
 *
 * @param {string[]} paths - Path segments
 * @return {string} Absolute path composed from `paths`
 */
export const resolve = vi.fn((...paths: string[]): string => {
  return path.resolve(...paths)
})
