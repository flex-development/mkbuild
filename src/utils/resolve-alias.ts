/**
 * @file Utilities - resolveAlias
 * @module mkbuild/utils/resolveAlias
 */

import type { Statement } from '#src/interfaces'
import * as pathe from 'pathe'

/**
 * Resolves a path alias in `statement.code` and `statement.specifier`.
 *
 * @param {Omit<Statement, 'type'>} statement - Statement object
 * @param {string} match - Possible path match
 * @param {string} source - Absolute path to source file
 * @return {void} Nothing when complete
 */
const resolveAlias = (
  statement: Omit<Statement, 'type'>,
  match: string,
  source: string
): void => {
  // do nothing if missing module specifier or path match
  if (!statement.specifier || !match) return

  // remove node_modules reference or set match to relative path from src
  match = /\/node_modules\//.test(match)
    ? match.replace(/.+\/node_modules\//, '')
    : pathe.relative(pathe.dirname(source), match).replace(/^(\w)/, './$1')

  // reset statement code, end index of code, and specifier
  statement.code = statement.code.replace(statement.specifier, match)
  statement.end = statement.start + statement.code.length
  statement.specifier = match
}

export default resolveAlias
