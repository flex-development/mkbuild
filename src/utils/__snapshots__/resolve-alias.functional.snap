// Vitest Snapshot v1

exports[`functional:utils/resolveAlias > resolve > node_modules > should resolve alias to module in node_modules directory 1`] = `
{
  "code": "import NodeEnv from '@flex-development/tutils/dist/enums/node-env'",
  "end": 66,
  "specifier": "@flex-development/tutils/dist/enums/node-env",
  "start": 0,
}
`;

exports[`functional:utils/resolveAlias > resolve > relative > should resolve alias to module in current directory 1`] = `
{
  "code": "import { foo } from './foobar'",
  "end": 30,
  "specifier": "./foobar",
  "start": 0,
}
`;

exports[`functional:utils/resolveAlias > resolve > relative > should resolve alias to module in different directory 1`] = `
{
  "code": "import type { Statement } from '../interfaces'",
  "end": 46,
  "specifier": "../interfaces",
  "start": 0,
}
`;
